Add code to create a HostOS ISO image based on CentOS packages

gyribeiro opened the pull request:
Refactor scripts that start actions, now there is a single script that supports subcommands that handle all required actions.
Support to create Host OS ISO image was added, the subcommand to perform this action was added, as well.

Commit:
Create python class to spin an ISO image.

Class MockPungiSpinner is created to spin a CentOS/Fedora image adding
HostOS packages, which uses mock and pungi (anaconda-based installation
spin tool) external tools.

Commit:
Add mock config file used in class MockPungiSpinner

farosas commented on lib/iso_spinner.py:
```diff
@@ -0,0 +1,128 @@
+    def __init__(self, config):
+        self.config = config
+        (self.distro, self.version, self.arch) = distro_utils.detect_distribution()
+        # only the major number must be consider
```
considered

gyribeiro replied:
Done

farosas commented on lib/iso_spinner.py:
```diff
@@ -0,0 +1,128 @@
+        self.config = config
+        (self.distro, self.version, self.arch) = distro_utils.detect_distribution()
+        # only the major number must be consider
+        self.version = re.sub("\..*$", "", self.version)
```
I'd go with `self.version.split('.')[0]` which is a bit more readable.

gyribeiro replied:
Done

farosas commented on lib/iso_spinner.py:
```diff
@@ -0,0 +1,128 @@
+
+    def _setup(self):
+        LOG.info("Initializing a chroot")
+        self._run_mock_command("-init")
```
Shouldn't it be '--init'?

gyribeiro replied:
Done

farosas commented on lib/iso_spinner.py:
```diff
@@ -0,0 +1,128 @@
+        # only the major number must be consider
+        self.version = re.sub("\..*$", "", self.version)
+
+    def _run_mock_command(self, cmd):
```
I might have missed this discussion, have we decided not to use the Mock class for this?

gyribeiro replied:
It was decided to keep rpm package builder and ISO image creator codes in separate modules, since they do not share any code.

Just a quick note: Mock class, in my opinion, is a misleading name because it is not a wrapper for mock command as you may notice, it is just rpm packages builder that uses mock.

mopsfelder commented on iso_config.yaml:
```diff
@@ -0,0 +1,37 @@
+rpm_dir: "./results"
+kickstart_file: "host-os.ks"
+hostos_package_group: "virtualization-hypervisor"
+package_group_list:
```
Do we need groups like desktop-debugging, dial-up, gnome-_, guest-desktop-agents, internet-_, multimedia, x11?

gyribeiro replied:
Short answer is not, the reason for this group list is that it came from official CentOS 7 ppc64le ISO, however, I've made it flexible exactly for this scenario, i.e., hostos decided to provide a custom CentOS ISO instead of the official ISO + hostos packages.

If the team prefers I can use a custom list instead of the official one. If this is the case we can discuss a minimum group list.

gyribeiro replied:
I did some tests and was able to reduce the package list to a minimum set, which speeds up the process and reduces the ISO size, hence, I will keep your suggestion and leave a tiny set of packages.

mopsfelder commented on iso_config.yaml:
```diff
@@ -0,0 +1,37 @@
+    dir: "/spin-repo"
+distro_repo_url:
+    base: "http://mirror.aus.stglabs.ibm.com/centos-altarch/7/os/ppc64le/"
+    updates: "http://mirror.aus.stglabs.ibm.com/centos-altarch/7/updates/ppc64le/"
```
This is an internal mirror.  What about http://mirror.centos.org/altarch/7/os/ppc64le/ for 'base' and http://mirror.centos.org/altarch/7/updates/ppc64le/ for 'updates'?

gyribeiro replied:
Agree, in the beginning it was using official CentOS mirror, but I've changed it during my tests, pointing them to an internal one to speed up the process and forgot to revert it.

olavphibm commented on iso_config.yaml:
```diff
@@ -0,0 +1,37 @@
+log_file: "/var/tmp/host-os-iso-creator.log"
```
In config.yaml, we have a "default" node containing all the other nodes. I think we should a similar convention in both files. That does not necessarily need to have the "default" node, but changing may be hard if we need to be backwards compatible.
The other config has `log_file: "/var/log/host-os/builds.log"` and it uses that same path for the three scripts that we have. I think we should use the "/var/log/host-os" here too and (unrelated to this PR) maybe use different files for different scripts (upgrade_versions, create_release_notes).

gyribeiro replied:
I didn't get your point regarding having or not  a "default" node in all yaml files. Do you want to add this node in the ISO yaml, as well?

Regarding the log, I was aware that the package builder uses /var/log/host-os to store its log, however, at the time I've written this code I did realized that I should run setup_environment.py previously to grant write access for /var/log. I will change it to "log_file: /var/log/host-os/iso.log", is it works for you?

olavphibm replied:
About the "default" node: I think we should follow the same format in both yaml file, but I don't really get why we have this "default" node, so up to you.
"/var/log/host-os/iso.log" is fine.

gyribeiro replied:
For now, I will keep it without the "default" node, I will start a discussion about whether it is worth to merge the config files or not.

olavphibm commented on iso_config.yaml:
```diff
@@ -0,0 +1,37 @@
+log_file: "/var/tmp/host-os-iso-creator.log"
+mock_config: "./extras/centOS/7.2/mock/spin-iso-epel-7-ppc64le.cfg"
+rpm_dir: "./results"
```
This could be a common config between this and the host-os-build "result_dir" RPM. It would be easier to generate the RPMs and then create the ISO in sequence. Maybe we could merge this file into config.yaml and use a sub-node "mock_iso" to separate the specific configs?

farosas replied:
+1 for merging and having a 'sub-node'

gyribeiro replied:
Initially, I've thought about it but as we decided to keep to rpm package builder and ISO creator code separate, I kept the same approach for the yaml file.

If the team agrees with this proposal, I don't see any problem in merger both yaml files as you suggest.

olavphibm commented on host-os-iso-creator.py:
```diff
@@ -0,0 +1,43 @@
+# Copyright (C) IBM Corp. 2016.
```
PEP8 is against using dashes in file names because they cannot be imported with a standard import statement. I suggest create_host_os_iso.py
I think we should rename the host-os-build.py file too, but I don't know if we can do this now with the backwards compatibility issue.

olavphibm replied:
About my backwards compatibility concerns, @maurorodrigues explained to me that this is a problem only to the versions repository, so they may be discarded here.
I keep my suggestion of naming to create_host_os_iso.py and (unrelated to this PR) create_host_os_build.py or create_host_os_packages.py

gyribeiro replied:
Ok, rename done ("-_ -> "_"), I've just used dash to follow the repository pattern.

olavphibm commented on host-os-iso-creator.py:
```diff
@@ -0,0 +1,43 @@
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--config-file', '-c',
+                        help='Path of the configuration file for create iso',
+                        default='./iso_config.yaml')
```
We need to find an easier way to have different arguments in different scripts. I don't have any suggestions right now though, so that's fine.

gyribeiro replied:
Agreed. We can have a discussion about it to find a better approach to avoid duplicated code spread in the project.

gyribeiro commented on lib/iso_spinner.py:
```diff
@@ -0,0 +1,128 @@
+        LOG.debug("Creating comps.xml")
+        # TODO: create comps.xml
+        comps_xml_file = "hostos-comps.xml"
+        self._run_mock_command("--copyin %s /" % comps_xml_file)
```
Done

gyribeiro commented on lib/iso_spinner.py:
```diff
@@ -0,0 +1,128 @@
+        mock_spin_repo_dir = self.config.get('mock_spin_repo').get('dir')
+        repo_url[mock_spin_repo_name] = "file://%s/" % mock_spin_repo_dir
+        with open(kickstart_file, "a+") as f:
+            for name, url in repo_url.items():
```
Done

gyribeiro commented on lib/iso_spinner.py:
```diff
@@ -0,0 +1,128 @@
+
+    def _spin(self):
+        LOG.info("Spinning ISO")
+        spin_cmd = "pungi -c %s --nosource --nodebuginfo --name %s --ver %s" % \
```
Done

olavphibm commented on lib/iso_spinner.py:
```diff
@@ -0,0 +1,131 @@
+        LOG.debug("Creating comps.xml")
+        # TODO: create comps.xml
+        comps_xml_file = "hostos-comps.xml"
+        comps_xml_chroot_path = "/"
```
I was actually thinking about setting
`comps_xml_chroot_path = os.path.join("/", comps_xml_file)`

gyribeiro replied:
done

olavphibm commented on lib/iso_spinner.py:
```diff
@@ -0,0 +1,131 @@
+                               (comps_xml_file, comps_xml_chroot_path))
+
+        LOG.debug("Creating spin repo")
+        createrepo_cmd = "createrepo -v -g /%s %s" % (comps_xml_file,
```
So we could use `comps_xml_chroot_path` here and make the link with the _run_mock_command parameters above more obvious to people unfamiliar with mock.

gutoyr commented on iso_config.yaml:
```diff
@@ -0,0 +1,37 @@
+    dir: "/spin-repo"
+distro_repo_url:
+    base: "http://mirror.aus.stglabs.ibm.com/centos-altarch/7/os/ppc64le/"
+    updates: "http://mirror.aus.stglabs.ibm.com/centos-altarch/7/updates/ppc64le/"
```
I did some tests and was able to reduce the package list to a minimum set, which speeds up the process and reduces the ISO size, hence, I will keep your suggestion and leave a tiny set of packages.

olavphibm commented on lib/iso_spinner.py:
```diff
@@ -0,0 +1,158 @@
+            # FIXME: linux-firmware provided by HostOS conflicts with
+            #        ivtv-firmware, this hack must be removed as soon as its
+            #        problem has been sorted out
+            f.write("-ivtv-firmware\n")
```
Do we need this here if it was already excluded from the repos above?

gyribeiro replied:
agree, removed

olavphibm commented on lib/iso_spinner.py:
```diff
@@ -0,0 +1,158 @@
+    def _save(self):
+        iso_file = "%s-DVD-%s-%s.iso" % (self.distro, self.arch, self.version)
+        checksum_file = ("%s-%s-%s-CHECKSUM" %
+                         (self.distro, self.version, self.arch))
```
I dislike that the ISO and checksum file names are non-uniform. If you're just copying those formats from CentOS, then that's fine.

gyribeiro replied:
This is odd, unfortunately we can not change it because pungi (spin tool) adopts different file name patterns for ISO and checksum.

Commit:
Rename host-os-build.py and change its interface to support subcommand

Rename host-os-build.py to host_os.py in order to make it in
compliance with PEP8 (dashes in file names should be avoided because
they cannot be imported with a standard import statement).

Added subcommand support to host_os.py interface, i.e.:
host_os.py [--args] <subcommand> [--subcommand-args]

Added build-package subcommand.

Commit:
Convert upgrade_versions.py script in a host_os subcommand

Commit:
Add "iso" node to config file (yaml)

Update config file, adding an "iso" node. Under this new node all setup
that will be used during the ISO spinning process

Commit:
Add suport to build-iso subcommand in host_os.py

Commit:
Convert create_release_notes.py script in a host_os subcommand

Commit:
Add python script to generate comps.xml content

Commit:
Convert setup_environment.py script in a host_os subcommand

olavphibm commented on host_os.py:
```diff
@@ -0,0 +1,38 @@
+              " directory for build scripts")
+        sys.exit(3)
+
+    sys.exit(SUBCOMMANDS.get(subcommand).run(CONF))
```
`get(subcommand)` will return None and raise an `AttributeError` if the key does not exist. I think it's more straightforward to use `[subcommand]`, so a `KeyError` will be raised.

gyribeiro replied:
Done.

farosas commented on lib/config.py:
```diff
@@ -98,56 +99,132 @@ def _setup_config_parser_options(self):
-        self.parser.add_argument('--repositories-path', '-R',
-                                 help='Directory where to clone code repositories',
-                                 default='/var/lib/host-os/repositories')
         self.parser.add_argument('--keep-builddir',
```
Isn't this specific to build-package?

gyribeiro replied:
Yes, moved it to build-package args

farosas commented on lib/config.py:
```diff
@@ -98,56 +99,132 @@ def _setup_config_parser_options(self):
+            help="Available subcommands")
+
+        # create the parser for the "build-package" subcommand
+        parser_package = subparser.add_parser('build-package',
```
What if we had a list of argument groups like so:
```
build_repo = [
    dict(option='--build-versions-repository-url', help='Build versions repository URL')
    dict(option='--build-versions-repo-dir', help='Directory to clone the build versions repository', default='./components')
]
push_repo = [...]
```

And then made each command 'subscribe' to them:
```
SUBCOMMANDS = [
    ('build-package', 'Build packages', [build_repo] ),
    ('release-notes', 'Create release notes', [push_repo, buld_repo]),
    ...
]
```

```
def add_args():
    for command, help, arg_groups in SUBCOMMANDS:
        subparser = subparser.add_parser(command, help)
        for arg_group in arg_groups:    
            for arg in arg_group:
                subparser.add_argument(**arg)
```

gyribeiro replied:
Done.

farosas commented on lib/config.py:
```diff
@@ -98,56 +99,132 @@ def _setup_config_parser_options(self):
+    def _add_build_repo_args(self, subparser):
+        """
+        Configures build repository arguments for subparser object to match the
+            expected configuration.
```
If my suggestion above is not taken, I think we should alter these docs to something more descriptive, like: "Args related to building the repo".

gyribeiro replied:
Not applicable after changes from previous comment.

farosas commented on lib/config.py:
```diff
@@ -98,56 +99,132 @@ def _setup_config_parser_options(self):
+        parser_iso = subparser.add_parser('build-iso', help="Build ISO image")
+        self._add_iso_args(parser_iso)
+
+    def _add_build_repo_args(self, subparser):
```
'subparser' here is inconsistent with the other methods' arguments. I'd prefer if we'd call all of them 'subparser', actually.

gyribeiro replied:
Not applicable after changes from previous comment.

farosas commented on host_os.py:
```diff
@@ -0,0 +1,50 @@
+from tools import build_package
+from tools import create_release_notes
+from tools import setup_environment
+from tools import upgrade_versions
```
I'm fine with `tools`, but just to leave it registered, I think another good option would be `lib/commands`.

gyribeiro replied:
ok

olavphibm commented on lib/config.py:
```diff
@@ -196,6 +203,16 @@ def _add_push_repo_args(self, subparser):
+        parser_setup_environment.add_argument('--user', '-u',
+                                              help='User login that will run '
+                                              'Host OS commands',
+                                              required=True)
```
The comment above on line breaking is also valid here.

gyribeiro replied:
Not applicable after last changes.

olavphibm commented on host_os.py:
```diff
@@ -19,24 +19,30 @@
+        sys.exit(3)
+
+    if os.getuid() is not 0 and subcommand == 'set-env':
+        print("%s should be run with root privileges" % " ".join(sys.argv))
```
You could simply print "The set-env command should be run with root privileges"

gyribeiro replied:
Done

gyribeiro commented on lib/iso_spinner.py:
```diff
@@ -0,0 +1,138 @@
+
+        LOG.debug("Creating spin repo directory")
+        mock_spin_repo_dir = self.config.get('mock_spin_repo').get('dir')
+        self._run_mock_command("--shell 'mkdir %s'" % mock_spin_repo_dir)
```
Done.

gyribeiro commented on lib/config.py:
```diff
@@ -136,6 +136,13 @@ def _setup_config_parser_options(self):
+        parser_setup_environment = subparser.add_parser('set-env',
+                                                        help="Setup user and "
+                                                        "directory for build "
+                                                        "scripts")
```
Not applicable after last changes.

gyribeiro commented on lib/config.py:
```diff
@@ -171,6 +160,30 @@ def _add_package_args(self, parser_package):
+        Configures the argument subparser release-notes object to match the
+        expected configuration.
+        """
+        parser_relase_notes.add_argument('--release-notes-repo-url',
```
Not applicable after last changes.

gyribeiro commented on tools/create_release_notes.py:
```diff
@@ -127,12 +125,11 @@ def publish_release_notes(
-    packages_names = (CONF.get('default').get('packages')
-                      or config.discover_packages())
+    packages_names = (CONF.get('default').get('packages') or
+                      config.discover_packages())
```
Done

gyribeiro commented on host_os.py:
```diff
@@ -19,24 +19,30 @@
         print("Please, do not run this script as root, run "
-              "setup_environment.py script in order to properly setup user and"
-              " directory for build scripts")
+              "host_os.py set-env --user <YOUR_USER_LOGIN> script in order to "
```
Done

olavphibm commented on lib/config.py:
```diff
@@ -24,6 +24,74 @@
+        dict(help='HTTP proxy URL'),
+}
+PACKAGE_ARGS = {
+    '--packages -p':
```
I'd prefer to use a tuple here. This way, we'll not need to split the string afterwards.
Tuples are immutable and hence can be used as dict keys.
`('--packages', '-p'): dict(...)`

olavphibm replied:
To make a tuple with a single element, leave a comma at the end: `('--keep-builddir',)`

gyribeiro replied:
Done.

olavphibm commented on lib/config.py:
```diff
@@ -101,59 +170,28 @@ def _setup_config_parser_options(self):
+            parser_command = subparser.add_parser(command, help=help_msg)
+            for arg_group in arg_groups:
+                for arg, options in arg_group.items():
+                    arg = arg.split()
```
If you follow my tuple suggestion, remove this line.
Also, rename to `args`, since there may be more than one.

gyribeiro replied:
Done.

Commit:
Always log external command outputs

mopsfelder commented on lib/iso_spinner.py:
```diff
@@ -0,0 +1,146 @@
+    def _run_mock_command(self, cmd):
+        try:
+            utils.run_command("mock -r %s %s" % (self.config.get('mock_config'),
+                                                 cmd))
```
Shouldn't we pass mock_args to this mock command as well?

gyribeiro replied:
Done

gyribeiro commented:
This closes #50
